phac-nml/irida-next
# multi omniauth models

## What does this PR do and why?

Allows signin authentication for SAML and Azure.
Using Omniauth through devise we can enable these authentication modes.

Documentation of omniauth providers being used
https://github.com/omniauth/omniauth
https://github.com/RIPAGlobal/omniauth-azure-activedirectory-v2
https://github.com/omniauth/omniauth-saml

## Changes
* add `uid` and `provider` to `user` table.
* Add `config/application.yml` configuration file via `figaro` gem https://github.com/laserlemon/figaro
  * this handles configuring the authentication providers
* omniauth callbacks handled by `app/controllers/users/omniauth_callbacks_controller.rb`
* callback controller handles user signin/creation via `app/models/user#from_omniauth`

## How to test locally

1. migrate db
2. edit `config/application.yml` to enable and configure omniauth providers
 * message me for certificate/client secret
 * Multiple omniauth providers can be enabled simultaneously but in production only one configuration will likely be used.
3. try signing in with the different providers
4. **Note: you must access irida next with `localhost` and not the local ip**

You should see this on the sign in screen
![image](https://github.com/phac-nml/irida-next/assets/16961365/590977e4-4b98-4dd0-ba36-4d8544ffcaf8)


### saml
If configured correctly, clicking "Sign in with SAML` will sign you in

### azure
If configured correctly, clicking "Sign in with AzureActivedirectoryV2` will sign you in

**Note: You will need to drop your db before testing this after testing saml, as they both use the same email and email is indexed and unique in the db schema**
`rake db:drop db:create db:migrate`

### developer
This provider is intended only for usage in development environments, so it's useful for testing different scenarios

1. Clicking "Sign in with Developer" will take you to a sign in page at `http://localhost:3000/users/auth/developer`

![image](https://github.com/phac-nml/irida-next/assets/16961365/afc0177c-03a5-4b24-bd79-e68426a4f608)

3. Enter a new name and email, click sign in.
4. Verify that you signed in.
5. Sign out, then sign in again via the same developer sign in page.
6. Verify that you are signed in again.
7. Sign out, then attempt to sign in with this account on the main sign in page at `http://localhost:3000/users/sign_in`
8. Verify that you are unable to sign in as this is not a local account.
9. Click the `Register now` button to go to the registration page, attempt to create an account with the email you used earlier.
10. Verify that you are unable to create this account.
11. Create a new local user via the registration page.
12. Try to sign in via Developer with the email used to create the new local user
13. Verify that the following error is displayed
![image](https://github.com/phac-nml/irida-next/assets/16961365/78316ea6-5b53-466a-92a6-dc81d4bb0e15)

**Note: the password reset route currently allows a user to change the randomly generated pw, and would allow them to sign in via the normal signin screen with their omniauth email. This will be addressed in a future PR**

Fixes #72
